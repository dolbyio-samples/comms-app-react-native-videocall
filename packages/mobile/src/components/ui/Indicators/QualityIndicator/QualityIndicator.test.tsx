import { defaultTheme as theme } from '@dolbyio/comms-uikit-common';
import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';

import { render, mergeStyles } from '../../../../utils/test-utils';

import QualityIndicator from './QualityIndicator';

const { colors } = theme;

const testID = 'testID';
const qualityLevel = 2;
const backgroundColor = 'grey.800';
const primaryColor = 'primary.500';
const secondaryColor = 'secondary.500';

describe('QualityIndicator component', () => {
  test('Passes TestID', () => {
    const { getByTestId } = render(<QualityIndicator testID={testID} qualityLevel={qualityLevel} />);
    expect(getByTestId(testID)).not.toBeNull();
  });
  test('Should color lines depending on quality', () => {
    const { getByTestId } = render(
      <QualityIndicator
        testID={testID}
        qualityLevel={qualityLevel}
        backgroundColor={backgroundColor}
        primaryColor={primaryColor}
        secondaryColor={secondaryColor}
      />,
    );
    const element1 = getByTestId('lineFirst');
    const element2 = getByTestId('lineSecond');
    const element3 = getByTestId('lineThird');
    const element4 = getByTestId('lineFourth');
    const element5 = getByTestId('lineFifth');
    const styles1 = mergeStyles(element1.props.style);
    const styles2 = mergeStyles(element2.props.style);
    const styles3 = mergeStyles(element3.props.style);
    const styles4 = mergeStyles(element4.props.style);
    const styles5 = mergeStyles(element5.props.style);
    expect(styles1.backgroundColor).toBe(qualityLevel >= 1 ? colors.secondary[500] : colors.primary[500]);
    expect(styles2.backgroundColor).toBe(qualityLevel >= 2 ? colors.secondary[500] : colors.primary[500]);
    expect(styles3.backgroundColor).toBe(qualityLevel >= 3 ? colors.secondary[500] : colors.primary[500]);
    expect(styles4.backgroundColor).toBe(qualityLevel >= 4 ? colors.secondary[500] : colors.primary[500]);
    expect(styles5.backgroundColor).toBe(qualityLevel >= 5 ? colors.secondary[500] : colors.primary[500]);
  });
  test('Should use whiteAlpha as backgroundColor in absence of backgroundColor prop', () => {
    const level = 5;
    const expStyles = {
      boxSizing: 'border-box',
      width: 26,
      height: 20,
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'flex-end',
      padding: 4,
      borderRadius: 6,
      backgroundColor: colors.whiteAlpha[500],
    };

    const { getByTestId } = render(
      <QualityIndicator
        testID={testID}
        qualityLevel={qualityLevel}
        primaryColor={primaryColor}
        secondaryColor={secondaryColor}
      />,
    );
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles).toMatchObject(expStyles);
  });
  test('Should use primaryColor as fill when qualityLevel is -1', () => {
    const qualityLevel = -1;
    const expStyles = {
      width: 2,
      height: 4,
      borderRadius: 2,
      backgroundColor: colors.primary[500],
    };
    const { getByTestId } = render(
      <QualityIndicator
        testID={testID}
        qualityLevel={qualityLevel}
        backgroundColor={backgroundColor}
        primaryColor={primaryColor}
        secondaryColor={secondaryColor}
      />,
    );
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles.backgroundColor).toBe(colors.grey[800]);

    const element1 = getByTestId('lineFirst');
    const element2 = getByTestId('lineSecond');
    const element3 = getByTestId('lineThird');
    const element4 = getByTestId('lineFourth');
    const element5 = getByTestId('lineFifth');
    const styles1 = mergeStyles(element1.props.style);
    const styles2 = mergeStyles(element2.props.style);
    const styles3 = mergeStyles(element3.props.style);
    const styles4 = mergeStyles(element4.props.style);
    const styles5 = mergeStyles(element5.props.style);

    expect(styles1).toMatchObject(expStyles);

    expStyles.height = 6;
    expect(styles2).toMatchObject(expStyles);

    expStyles.height = 8;
    expect(styles3).toMatchObject(expStyles);

    expStyles.height = 10;
    expect(styles4).toMatchObject(expStyles);

    expStyles.height = 12;
    expect(styles5).toMatchObject(expStyles);
  });
  test('Should use default primaryColor as fill when primaryColor is not set', () => {
    const qualityLevel = 3;
    const { getByTestId } = render(
      <QualityIndicator
        testID={testID}
        qualityLevel={qualityLevel}
        backgroundColor={backgroundColor}
        secondaryColor={secondaryColor}
      />,
    );
    const element1 = getByTestId('lineFirst');
    const element2 = getByTestId('lineSecond');
    const element3 = getByTestId('lineThird');
    const element4 = getByTestId('lineFourth');
    const element5 = getByTestId('lineFifth');
    const styles1 = mergeStyles(element1.props.style);
    const styles2 = mergeStyles(element2.props.style);
    const styles3 = mergeStyles(element3.props.style);
    const styles4 = mergeStyles(element4.props.style);
    const styles5 = mergeStyles(element5.props.style);

    expect(styles1.backgroundColor).toBe(qualityLevel >= 1 ? colors.secondary[500] : colors.grey[600]);
    expect(styles2.backgroundColor).toBe(qualityLevel >= 2 ? colors.secondary[500] : colors.grey[600]);
    expect(styles3.backgroundColor).toBe(qualityLevel >= 3 ? colors.secondary[500] : colors.grey[600]);
    expect(styles4.backgroundColor).toBe(qualityLevel >= 4 ? colors.secondary[500] : colors.grey[600]);
    expect(styles5.backgroundColor).toBe(qualityLevel >= 5 ? colors.secondary[500] : colors.grey[600]);
  });
  test('Should use default secondaryColor as fill when secondaryColor is not set', () => {
    const qualityLevel = 5;
    const { getByTestId } = render(
      <QualityIndicator testID={testID} qualityLevel={qualityLevel} backgroundColor={backgroundColor} />,
    );
    const element1 = getByTestId('lineFirst');
    const element2 = getByTestId('lineSecond');
    const element3 = getByTestId('lineThird');
    const element4 = getByTestId('lineFourth');
    const element5 = getByTestId('lineFifth');
    const styles1 = mergeStyles(element1.props.style);
    const styles2 = mergeStyles(element2.props.style);
    const styles3 = mergeStyles(element3.props.style);
    const styles4 = mergeStyles(element4.props.style);
    const styles5 = mergeStyles(element5.props.style);

    expect(styles1.backgroundColor).toBe(qualityLevel >= 1 ? colors.white : colors.grey[600]);
    expect(styles2.backgroundColor).toBe(qualityLevel >= 2 ? colors.white : colors.grey[600]);
    expect(styles3.backgroundColor).toBe(qualityLevel >= 3 ? colors.white : colors.grey[600]);
    expect(styles4.backgroundColor).toBe(qualityLevel >= 4 ? colors.white : colors.grey[600]);
    expect(styles5.backgroundColor).toBe(qualityLevel >= 5 ? colors.white : colors.grey[600]);
  });
});

describe('QualityIndicator component snapshot', () => {
  test('Should match when no optional parameters are specified', () => {
    const qualityLevel = 3;
    const tree = renderer.create(<QualityIndicator testID={testID} qualityLevel={qualityLevel} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Should match when all parameters are specified', () => {
    const qualityLevel = 3;
    const tree = renderer
      .create(
        <QualityIndicator
          testID={testID}
          qualityLevel={qualityLevel}
          backgroundColor={backgroundColor}
          primaryColor={primaryColor}
          secondaryColor={secondaryColor}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
