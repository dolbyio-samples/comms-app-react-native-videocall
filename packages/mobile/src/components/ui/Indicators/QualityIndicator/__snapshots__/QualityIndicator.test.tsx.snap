// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QualityIndicator component snapshot Should match when all parameters are specified 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "borderRadius": 6,
        "boxSizing": "border-box",
        "flexDirection": "row",
        "height": 20,
        "justifyContent": "space-between",
        "padding": 4,
        "width": 26,
      },
      Object {
        "backgroundColor": "#292930",
      },
    ]
  }
  testID="testID"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 4,
        },
        Object {
          "backgroundColor": "#628EF4",
        },
      ]
    }
    testID="lineFirst"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 6,
        },
        Object {
          "backgroundColor": "#628EF4",
        },
      ]
    }
    testID="lineSecond"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 8,
        },
        Object {
          "backgroundColor": "#628EF4",
        },
      ]
    }
    testID="lineThird"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 10,
        },
        Object {
          "backgroundColor": "#8829CC",
        },
      ]
    }
    testID="lineFourth"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 12,
        },
        Object {
          "backgroundColor": "#8829CC",
        },
      ]
    }
    testID="lineFifth"
  />
</View>
`;

exports[`QualityIndicator component snapshot Should match when no optional parameters are specified 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "borderRadius": 6,
        "boxSizing": "border-box",
        "flexDirection": "row",
        "height": 20,
        "justifyContent": "space-between",
        "padding": 4,
        "width": 26,
      },
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0.36)",
      },
    ]
  }
  testID="testID"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 4,
        },
        Object {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
    testID="lineFirst"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 6,
        },
        Object {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
    testID="lineSecond"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 8,
        },
        Object {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
    testID="lineThird"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 10,
        },
        Object {
          "backgroundColor": "#535359",
        },
      ]
    }
    testID="lineFourth"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 2,
          "width": 2,
        },
        Object {
          "height": 12,
        },
        Object {
          "backgroundColor": "#535359",
        },
      ]
    }
    testID="lineFifth"
  />
</View>
`;
